README.md
# CodeWall

[![GitHub Workflow Status](https://img.shields.io/github/workflow/status/abhinavk99/codewall/CodeWall%20CI?style=flat-square)](https://github.com/abhinavk99/codewall/actions)
[![Visual Studio Marketplace Version](https://img.shields.io/visual-studio-marketplace/v/abhinavk99.codewall?style=flat-square)](https://marketplace.visualstudio.com/items?itemName=abhinavk99.codewall)
[![Visual Studio Marketplace Installs](https://img.shields.io/visual-studio-marketplace/i/abhinavk99.codewall?style=flat-square)](https://marketplace.visualstudio.com/items?itemName=abhinavk99.codewall)
[![Visual Studio Marketplace Downloads](https://img.shields.io/visual-studio-marketplace/d/abhinavk99.codewall?style=flat-square)](https://marketplace.visualstudio.com/items?itemName=abhinavk99.codewall)
[![code style: prettier](https://img.shields.io/badge/code_style-prettier-ff69b4.svg?style=flat-square)](https://github.com/prettier/prettier)
[![GitHub Repo stars](https://img.shields.io/github/stars/abhinavk99/codewall?style=flat-square)](https://github.com/abhinavk99/codewall)
[![GitHub](https://img.shields.io/github/license/abhinavk99/codewall?style=flat-square)](https://github.com/abhinavk99/codewall/blob/master/LICENSE)
[![PRs Welcome](https://img.shields.io/badge/PRs-welcome-brightgreen.svg?style=flat-square)](http://makeapullrequest.com)

VS Code extension that checks if code goes past your rulers. Install it [here](https://marketplace.visualstudio.com/items?itemName=abhinavk99.codewall)!

## Features

Shows warnings at the specific locations where code goes past your rulers.

![](https://raw.githubusercontent.com/abhinavk99/codewall/master/images/window.PNG)

![](https://raw.githubusercontent.com/abhinavk99/codewall/master/images/message.PNG)

## Settings

This extension has the following settings:

- `codewall.openProblemsPane`: Determines whether to open the problems pane to show warnings when lines cross rulers. Is true by default.

## Usage

Add a rulers using the `editor.rulers` setting (`Command Palette -> Preferences: Open Settings (JSON)`).

Valid Values:

```json
"editor.rulers": [
  {"column": 90, "color": "#000000"},
  {"column": 105, "color": "#ff0000"}
]
```

```json
"editor.rulers": [90, 105]
```

## CodeWall for Atom

Don't use VS Code? Try [CodeWall](https://github.com/Oceanwall/CodeWall) for Atom, made by Oceanwall!


CHANGELOG.md
# Changelog

## [1.2.2] - 2021-01-14

### Changed

- Switched from checking lines when getting onDidSaveTextDocument to onWillSaveTextDocument event for minor performance boost
- Refactored code

## [1.2.1] - 2021-01-12

### Added

- README badges
- Dev
  - Github Actions CI testing

## [1.2.0] - 2021-01-12

### Added

- Added support for ruler objects from configuration (e.g. {"column": 90, "color": "#000"})
- Added usage notes in README
- Dev
  - Switched from Yarn to NPM
  - Added and used linter and formatter

### Changed

- Changed extension activation event from `*` to `onStartupFinished` so it'll start after VS Code is done starting
- Ignore rulers that have a column number of <= 0
- Performance optimizations
  - Switched from linear search to binary search for determining whether a line crosses one of the rulers
  - Store rulers configuration and update it when configuration gets changed to avoid repetitive VS Code API calls
- Dev
  - Revamped test runner code and added unit tests

### Security

- Dev
  - Updated dev dependencies

## [1.1.0] - 2019-10-28

### Added

- Added setting `codewall.openProblemsPane` for whether to open it to show warnings

## [1.0.7] - 2019-07-22

### Changed

- Updated dev dependencies

## [1.0.6] - 2019-07-22

### Fixed

- Fixed repository link in package.json

## [1.0.5] - 2019-06-15

### Changed

- Updated dev dependencies

## [1.0.4] - 2018-03-09

### Changed

- Don't show the problems pane if there are no problems in the file

## [1.0.3] - 2018-11-16

### Fixed

- Added missing semicolon

## [1.0.2] - 2018-11-16

### Fixed

- Fixed images in README

## [1.0.1] - 2018-11-16

### Fixed

- Fixed description in package.json

## [1.0.0] - 2018-11-16

### Added

- Initial release


src/test/suite/extension.test.ts
import * as assert from 'assert';

// You can import and use all API from the 'vscode' module
// as well as import your extension to test it
import vscode from 'vscode';
import * as codewall from '../../extension';

interface TestCase {
  expectedResult: number;
  message: string;
}

interface NumberTestCase extends TestCase {
  input: number;
}

interface RulerTestCase extends TestCase {
  ruler: codewall.Ruler;
}

interface RulersTestCase extends TestCase {
  ruler1: codewall.Ruler;
  ruler2: codewall.Ruler;
}

suite('CodeWall Test Suite', () => {
  vscode.window.showInformationMessage('Start all CodeWall tests.');
  const codeWall = new codewall.CodeWall();
  const rulers = [
    {
      column: 90,
      color: '#000000',
    },
    100,
    {
      column: 120,
      color: '#ff0000',
    },
  ];

  const rulerComparatorTestcases = [
    {
      ruler1: 5,
      ruler2: 8,
      expectedResult: -3,
      message: 'Get negative number when number ruler a is before number ruler b',
    },
    {
      ruler1: 34,
      ruler2: 18,
      expectedResult: 16,
      message: 'Get positive number when number ruler a is after number ruler b',
    },
    {
      ruler1: 3,
      ruler2: 3,
      expectedResult: 0,
      message: 'Get 0 when number ruler a is at same position as number ruler b',
    },
    {
      ruler1: { column: 5, color: '#000000' },
      ruler2: 8,
      expectedResult: -3,
      message: 'Get negative number when object ruler a is before number ruler b',
    },
    {
      ruler1: 34,
      ruler2: { column: 18, color: '#000000' },
      expectedResult: 16,
      message: 'Get positive number when number ruler a is after object ruler b',
    },
    {
      ruler1: { column: 3, color: '#000000' },
      ruler2: { column: 3, color: '#000000' },
      expectedResult: 0,
      message: 'Get 0 when object ruler a is at same position as object ruler b',
    },
  ];
  rulerComparatorTestcases.forEach((testCase: RulersTestCase) => {
    test(`${testCase.message}`, () => {
      assert.strictEqual(codeWall.rulerComparator(testCase.ruler1, testCase.ruler2), testCase.expectedResult);
    });
  });

  const getRulerColumnNumberTestCases = [
    { ruler: 5, expectedResult: 5, message: 'Get ruler column number from number' },
    { ruler: { column: 5, color: '#000000' }, expectedResult: 5, message: 'Get ruler column number from object' },
  ];
  getRulerColumnNumberTestCases.forEach((testCase: RulerTestCase) => {
    test(`${testCase.message}`, () => {
      assert.strictEqual(codeWall.getRulerColumnNumber(testCase.ruler), testCase.expectedResult);
    });
  });

  const getRulerColumnNumberThatLineCrossedTestCases = [
    { input: 80, expectedResult: -1, message: 'Get -1 when line does not cross any rulers' },
    { input: 95, expectedResult: 90, message: 'Get first ruler column number when line crosses first ruler' },
    { input: 110, expectedResult: 100, message: 'Get second ruler column number when line crosses second ruler' },
    { input: 150, expectedResult: 120, message: 'Get last ruler column number when line crosses last ruler' },
    { input: 90, expectedResult: -1, message: 'Get -1 when line ends at first ruler' },
    { input: 100, expectedResult: 90, message: 'Get first ruler column number when line ends at second ruler' },
    { input: 120, expectedResult: 100, message: 'Get second ruler column number when line ends at third ruler' },
  ];
  getRulerColumnNumberThatLineCrossedTestCases.forEach((testCase: NumberTestCase) => {
    test(`${testCase.message}`, () => {
      assert.strictEqual(codeWall.getRulerColumnNumberThatLineCrossed(testCase.input, rulers), testCase.expectedResult);
    });
  });
});


src/extension.ts
'use strict';

import vscode from 'vscode';
import binarySearch from 'binary-search';

// A ruler in VS Code settings can either be a number or an object mapping strings to numbers/strings
// e.g. 90
// e.g. {"column": 90, "color": "#000"}
export type Ruler = number | RulerObject;

export interface RulerObject {
  column: number;
  color: string;
}

export function activate(context: vscode.ExtensionContext): void {
  const codeWall = new CodeWall();

  if (vscode.window.activeTextEditor != undefined) {
    codeWall.checkCrossingWall(vscode.window.activeTextEditor.document);
  }

  context.subscriptions.push(
    vscode.window.onDidChangeActiveTextEditor((editor: vscode.TextEditor | undefined) => {
      if (editor != undefined) {
        codeWall.checkCrossingWall(editor.document);
      }
    }),
    vscode.workspace.onWillSaveTextDocument((textDocumentWillSaveEvent: vscode.TextDocumentWillSaveEvent) => {
      codeWall.checkCrossingWall(textDocumentWillSaveEvent.document);
    }),
    vscode.workspace.onDidChangeConfiguration((configChangeEvent: vscode.ConfigurationChangeEvent) => {
      if (configChangeEvent.affectsConfiguration(CodeWall.EDITOR_CONFIG)) {
        codeWall.setRulers();
      }
      if (configChangeEvent.affectsConfiguration(CodeWall.CODEWALL_CONFIG)) {
        codeWall.setOpenProblemsPane();
      }
    }),
    codeWall
  );
}

export function deactivate(): void {
  return;
}

export class CodeWall {
  static readonly EDITOR_CONFIG = 'editor';
  static readonly CODEWALL_CONFIG = 'codewall';

  readonly diagnosticCollection: vscode.DiagnosticCollection;

  rulers: Ruler[] = [];
  openProblemsPane: boolean;

  constructor() {
    this.diagnosticCollection = vscode.languages.createDiagnosticCollection('CodeWall');
    this.setRulers();
    this.openProblemsPane = false;
    this.setOpenProblemsPane();
  }

  /**
   * Gets and sets all the rulers in the VS Code settings in ascending column number order ignoring rulers at column <= 0.
   */
  public setRulers(): void {
    this.rulers = vscode.workspace.getConfiguration(CodeWall.EDITOR_CONFIG).get('rulers', []);
    this.rulers.sort(this.rulerComparator.bind(this));
    this.rulers.filter((ruler) => this.getRulerColumnNumber(ruler) > 0);
  }

  public setOpenProblemsPane(): void {
    this.openProblemsPane =
      vscode.workspace.getConfiguration(CodeWall.CODEWALL_CONFIG).get('openProblemsPane') ?? false;
  }

  /**
   * Goes through the document and adds warnings for any lines crossing any rulers.
   * @param document VS Code text document
   */
  public checkCrossingWall(document: vscode.TextDocument): void {
    this.diagnosticCollection.clear();
    const diagnostics: vscode.Diagnostic[] = [];

    if (this.rulers.length == 0) {
      return;
    }

    for (const line of this.getLines(document)) {
      const crossedRulerColumnNumber = this.getRulerColumnNumberThatLineCrossed(line.range.end.character, this.rulers);
      if (crossedRulerColumnNumber != -1) {
        this.addDiagnosticWarningToDiagnostics(line, crossedRulerColumnNumber, diagnostics);
      }
    }

    if (diagnostics.length > 0) {
      this.diagnosticCollection.set(document.uri, diagnostics);

      // Open problems pane if setting is on
      if (this.openProblemsPane) {
        vscode.commands.executeCommand('workbench.action.problems.focus');
      }
    }
  }

  /**
   * Comparator function for sorting rulers in ascending order based on column number.
   * @param a Ruler 1
   * @param b Ruler 2
   * @returns Negative number if a is before b, Positive number if a is after b, 0 if equal
   */
  public rulerComparator(a: Ruler, b: Ruler): number {
    return this.getRulerColumnNumber(a) - this.getRulerColumnNumber(b);
  }

  /**
   * Since a ruler can either be a number or an object, this method abstracts out getting
   * the column number from the ruler.
   * @param ruler VS Code ruler from settings
   * @returns Ruler column number
   */
  public getRulerColumnNumber(ruler: Ruler): number {
    if (typeof ruler === 'number') {
      return ruler;
    }
    return ruler.column;
  }

  /**
   * Gets all the lines in the text document as an iterator.
   * @param document VS Code text document
   * @returns Iterator of lines in the text document
   */
  private *getLines(document: vscode.TextDocument) {
    const lineCount = document.lineCount;
    for (let lineNumber = 0; lineNumber < lineCount; lineNumber++) {
      yield document.lineAt(lineNumber);
    }
  }

  /**
   * Gets the column number of the ruler the line passed or -1.
   * @param lineEndColumnNumber Column number of the last character in the line
   * @param rulers List of rulers in ascending order
   * @returns The column number of the ruler the line passed or -1
   */
  public getRulerColumnNumberThatLineCrossed(lineEndColumnNumber: number, rulers: Ruler[]): number {
    let insertionIndex = binarySearch(rulers, lineEndColumnNumber, this.rulerComparator.bind(this));
    if (insertionIndex < 0) {
      // -(index + 1) is returned by binarySearch when the value is not found in the array, where index is where the value
      // would be inserted to maintain sorted order, so we do simple math to extract the index from what was returned.
      insertionIndex = -insertionIndex - 1;
    }
    if (insertionIndex == 0) {
      // Line ends before any rulers
      return -1;
    }
    return this.getRulerColumnNumber(rulers[insertionIndex - 1]);
  }

  /**
   * Adds VS Code Diagnostic warning to list of Diagnostic objects.
   * @param line Line that crosses a ruler
   * @param crossedRulerColumnNumber Column number of the ruler being crossed
   * @param diagnostics List of Diagnostic objects being added to
   */
  private addDiagnosticWarningToDiagnostics(
    line: vscode.TextLine,
    crossedRulerColumnNumber: number,
    diagnostics: vscode.Diagnostic[]
  ) {
    diagnostics.push({
      code: '',
      message: `Line ${line.lineNumber + 1} is longer than ruler at column ${crossedRulerColumnNumber}.`,
      range: new vscode.Range(new vscode.Position(line.lineNumber, crossedRulerColumnNumber - 1), line.range.end),
      severity: vscode.DiagnosticSeverity.Warning,
    });
  }

  public dispose(): void {
    this.diagnosticCollection.dispose();
  }
}
